libsolidity/ast/Types.h

```diff
@@ -252,7 +252,10 @@ class Type
 	/// @returns an upper bound on the total storage size required by this type, descending
 	/// into structs and statically-sized arrays. This is mainly to ensure that the storage
 	/// slot allocation algorithm does not overflow, it is not a protection against collisions.
-	virtual bigint storageSizeUpperBound() const { return 1; }
+	bigint storageSizeUpperBound() const { return sizeUpperBound(DataLocation::Storage); }
+	/// @returns an upper bound on the total size required by this type
+	/// for the given location (in slots for storage, in bytes otherwise)
+	virtual bigint sizeUpperBound(DataLocation _loc) const { return (_loc == DataLocation::Storage) ? 1 : memoryHeadSize(); }
```

Comment:
As noted in another comment, today this actually looks like a *lower* bound to me :-D.
Especially in the presence of dynamically sized types, actual instances of the type may very well be larger.
It may also be the exact size of the static part of the type or something like that...

When are we actually *over*estimating right now?

Reply:
I think it is more like _conservative estimation of the lower bound_ :). We do not pack smaller types in slots, right? Just `roughSize(DataLocation)` with a comment maybe?